### Cardiac Arrhythmia Multy-Class Classification 

Analyze data and address missing data if there is any. 

Decide aboute a good evaluation strategy and justify your choice. 

Find the best parameters for the following classification models: 
- KNN classifcation 
- Logistic Regression
- Linear Supprt Vector Machine
- Kerenilzed Support Vector Machine
- Decision Tree
- Random Forest 

Then use different bagging and boosting methods to boost the results? Do you see any significant change? Why or why not? 

Next, use data reduction method you have learned in class to reduce the size of data, and agian try above models. Do you get better results? Justify your answer. 

<font color = 'red'>Due date for full credit: April 4, 11:59 PM
    <br>
    Due date for partial credit: April 6, 11:59 PM.
    <br> No submission will be accepted after April 6. 
    <br> Please note that your term paper is also due April 6. 
    </font>

### Importing Packages

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.svm import LinearSVC
from sklearn.metrics import confusion_matrix

### Importing Data

In [2]:
Data=pd.read_csv("cardiac_arrhythmia.csv",header=None)
data=Data # Creating Duplicate for the data
data

0    1    2    3    4    5    6    7    8    9   ...   270   271   272  \
0     75    0  190   80   91  193  371  174  121  -16 ...   0.0   9.0  -0.9   
1     56    1  165   64   81  174  401  149   39   25 ...   0.0   8.5   0.0   
2     54    0  172   95  138  163  386  185  102   96 ...   0.0   9.5  -2.4   
3     55    0  175   94  100  202  380  179  143   28 ...   0.0  12.2  -2.2   
4     75    0  190   80   88  181  360  177  103  -16 ...   0.0  13.1  -3.6   
5     13    0  169   51  100  167  321  174   91  107 ...  -0.6  12.2  -2.8   
6     40    1  160   52   77  129  377  133   77   77 ...   0.0   6.5   0.0   
7     49    1  162   54   78    0  376  157   70   67 ...   0.0   8.2  -1.9   
8     44    0  168   56   84  118  354  160   63   61 ...   0.0   7.0  -1.3   
9     50    1  167   67   89  130  383  156   73   85 ...  -0.6  10.8  -1.7   
10    62    0  170   72  102  135  401  156   83   72 ...  -0.5   9.0  -2.0   
11    45    1  165   86   77  143  373  150   65   12 ...   0.0   4.4  -2.2   
12    54    1  172   58   78  155  382  163   81  -24 ...   0.0   6.3  -2.1   
13    30    0  170   73   91  180  355  157  104   68 ...  -0.9  12.3   0.0   
14    44    1  160   88   77  158  399  163   94   46 ...  -0.6  12.4   0.0   
15    47    1  150   48   75  132  350  169   65   36 ...   0.0   7.7  -0.8   
16    47    0  171   59   82  145  347  169   61   77 ...   0.0   9.4  -1.7   
17    46    1  158   58   70  120  353  122   52   57 ...   0.0   6.6   0.0   
18    73    0  165   63   91  154  392  175   83   73 ...   0.0   5.7   0.0   
19    57    1  166   72   82  181  399  158   79  -12 ...   0.0   7.7  -0.9   
20    28    1  160   58   83  251  383  189  183   50 ...  -0.6   9.1  -1.4   
21    45    0  169   67   90  122  336  177   78   81 ...  -0.6   8.3  -1.8   
22    36    1  153   75   71  132  364  169   82   62 ...   0.0   8.9  -1.0   
23    57    1  165   59   75  157  406  143   92    4 ...   0.0   6.7  -0.5   
24    40    1  153   55   82  140  388  149   82   52 ...   0.0  13.6   0.0   
25    44    0  169   80  109  128  382  195   60  -34 ...   0.0   6.9   0.0   
26    34    0  170   73   94  186  373  224  125   90 ...   0.0  15.3  -1.1   
27    31    1  160   54   95  161  407  168   83   10 ...   0.0  12.7  -1.8   
28    56    1  164   65   90  164  420  381   99   -8 ...   0.0   5.4   0.0   
29    51    1  160   83   96  147  400  301   82  -37 ...   0.0   7.3  -3.9   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ... ...   ...   ...   ...   
422   29    1  162   57   83  164  359  154   69   64 ...   0.0  14.1  -2.2   
423   51    0  186   95   94  203  367  171  106   -7 ...   0.0   9.6  -3.5   
424    7    0  119   21  140  157  438  226   81  -40 ...   0.0  10.0  -2.1   
425   36    0  171   93   87  150  362  177   96   44 ...   0.0  10.3  -0.8   
426   35    1  160   53   55  163  340  162  102   40 ...   0.0   8.7  -0.5   
427   58    0  160   65  133  148  417  260   92 -158 ...  -0.4   6.4  -3.5   
428   64    0  160   63   83    0  364  120   90   29 ...   0.0   6.7  -0.4   
429    8    1  130   24   77  125  358  159   70   87 ...   0.0  11.3  -2.1   
430   11    0  138   29  123  145  361  221   80  112 ...  -3.4  19.6  -4.2   
431   47    0  166   56   79  145  381  173  101   52 ...   0.0   8.5   0.0   
432   11    0  140   42   88  123  362  228   81  -18 ...   0.0  17.1  -7.1   
433   70    0  167   60   80  149  290  128   93  -67 ...   0.0   2.7  -5.4   
434   20    0  178   65   88  155  360  163   71  -22 ...  -0.5  10.2   0.0   
435   39    1  164   62   79  155  367  153   95   50 ...   0.0   9.7  -0.7   
436   32    1  164   57   77  144  340  148   82   27 ...  -0.6   9.9  -0.6   
437   35    1  155   63   87  142  391  137   88   66 ...   0.0  10.7   0.0   
438   37    0  175   82   88  146  357  179   72    1 ...  -0.4  13.5  -1.2   
439   49    1  168   66   94  170  383  152  115   92 ...   0.0   8.2  -0.7   
440   37    0  176   72   88  153  389  172   89   67 .

##### There are 452 Rows and 280 columns in the data. We can say total number of observations are very less when compared to number of features.

### Missing values in tha Data"

In [3]:
data[13].head()

0     ?
1     ?
2    23
3     ?
4     ?
Name: 13, dtype: object

##### There are missing values in tha data but we can find them as "?" instead of "nan"

### Dropping Null values and Replacing the Missing Values with 0

In [4]:
data=data.dropna()
for i in data.columns:
    data[i]=data[i].replace("?",0)
data[13].head()

0     0
1     0
2    23
3     0
4     0
Name: 13, dtype: object

In [5]:
data.describe()

0           1           2           3           4           5    \
count  452.000000  452.000000  452.000000  452.000000  452.000000  452.000000   
mean    46.471239    0.550885  166.188053   68.170354   88.920354  155.152655   
std     16.466631    0.497955   37.170340   16.590803   15.364394   44.842283   
min      0.000000    0.000000  105.000000    6.000000   55.000000    0.000000   
25%     36.000000    0.000000  160.000000   59.000000   80.000000  142.000000   
50%     47.000000    1.000000  164.000000   68.000000   86.000000  157.000000   
75%     58.000000    1.000000  170.000000   79.000000   94.000000  175.000000   
max     83.000000    1.000000  780.000000  176.000000  188.000000  524.000000   

              6           7           8           9       ...             270  \
count  452.000000  452.000000  452.000000  452.000000     ...      452.000000   
mean   367.207965  169.949115   90.004425   33.676991     ...       -0.278982   
std     33.385421   35.633072   25.826643   45.431434     ...        0.548876   
min    232.000000  108.000000    0.000000 -172.000000     ...       -4.100000   
25%    350.000000  148.000000   79.000000    3.750000     ...       -0.425000   
50%    367.000000  162.000000   91.000000   40.000000     ...        0.000000   
75%    384.000000  179.000000  102.000000   66.000000     ...        0.000000   
max    509.000000  381.000000  205.000000  169.000000     ...        0.000000   

              271         272         273    274         275         276  \
count  452.000000  452.000000  452.000000  452.0  452.000000  452.000000   
mean     9.048009   -1.457301    0.003982    0.0    0.514823    1.222345   
std      3.472862    2.002430    0.050118    0.0    0.347531    1.426052   
min      0.000000  -28.600000    0.000000    0.0   -0.800000   -6.000000   
25%      6.600000   -2.100000    0.000000    0.0    0.400000    0.500000   
50%      8.800000   -1.100000    0.000000    0.0    0.500000    1.350000   
75%     11.200000    0.000000    0.000000    0.0    0.700000    2.100000   
max     23.600000    0.000000    0.800000    0.0    2.400000    6.000000   

              277         278         279  
count  452.000000  452.000000  452.000000  
mean    19.326106   29.473230    3.880531  
std     13.503922   18.493927    4.407097  
min    -44.200000  -38.600000    1.000000  
25%     11.450000   17.550000    1.000000  
50%     18.100000   27.900000    1.000000  
75%     25.825000   41.125000    6.000000  
max     88.800000  115.900000   16.000000  

[8 rows x 275 columns]

### Splitting the Target variable from the Data.
#### Input Variable = x 
#### Target variable = y

In [6]:
x=data.iloc[:,:-1]
y=data.iloc[:,-1]

### Generating the Target variable for MultipleClass Classiffication and for Detectecting Arrhythmia
#### y is for Multiclass Classification
#### z is for detecting Arrhythmia

In [7]:
print(y.value_counts())
z = y!=1
print(z.value_counts())

1     245
10     50
2      44
6      25
16     22
4      15
3      15
5      13
9       9
15      5
14      4
7       3
8       2
Name: 279, dtype: int64
False    245
True     207
Name: 279, dtype: int64


#### There is NO class imbalence problem in the dataset.
#### There are 207 cases with arrhythmia and 245 records with no arrhythmia

In the Target variable, there are number of instances related to a person having no Arrhythmia with target label "1" and a person with symptoms of Arrhythmia with labels ranging from "2 to 15".

In the dataset, we have 245 observations of normal person and remaining 207 observations are spread across 14 labelled data. So, there is huge class imbalance problem, when we consider multi-class data.

Because of the above reason, I classified the target variable as two categories.
1. Obsevation with no symptoms
2. Observations with symptoms.

By doing this, we can eliminate the class imbalance problem. If the prediction says that an observations is having symptoms, then we can do the predictions on multi-class data to know the related symptoms.

### Splitting the data for Training and Testing

In [8]:
x_train, x_test, y_train, y_test= train_test_split(x,y,random_state=0)
x_train, x_test, z_train, z_test= train_test_split(x,z,random_state=0)
print("Number of rows in training", x_train[0].count())
print("Number of rows in testing", x_test[0].count())

Number of rows in training 339
Number of rows in testing 113


### Scalling the Data

In [9]:
scaler = MinMaxScaler()
x_trainscaled = scaler.fit_transform(x_train)
x_testscaled = scaler.transform(x_test)

Scaling the data is very important to increase the speed of learning rate. Initially the data is spread across different ranges forming eliptical curves during the optimization of the gradient decent. By applying the scaling function we can get all the columns in a simillar range which helps the gradient decent to optimize the values of the coefficient in a very few number of steps and eliminate the changes of strucking in a local minima.

### KNN

#### KNN with Multiclass Data

In [10]:
knnclas=KNeighborsClassifier()
k_value={'n_neighbors':range(1,100)}
grid=GridSearchCV(knnclas,param_grid=k_value)
grid.fit(x_trainscaled,y_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,y_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,y_test))
print(grid.best_params_)
print(grid.best_estimator_)
knnclasp=grid.predict(x_testscaled)
confusion_matrix(y_test,knnclasp)

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


The best training score is :  0.7080
The best test score is :  0.6106
{'n_neighbors': 2}
KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=2, p=2,
           weights='uniform')


array([[58,  1,  0,  1,  1,  0,  0,  0,  0,  0,  0],
       [ 8,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  4,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0],
       [ 2,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [10,  1,  0,  0,  1,  0,  0,  0,  1,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0]])

In [11]:
knnm=KNeighborsClassifier(n_neighbors=2)
knn=knnm.fit(x_trainscaled,y_train)
print("knn Training scaore", knn.score(x_trainscaled,y_train))
print("knn Test score",knn.score(x_testscaled,y_test))
knnp=knn.predict(x_testscaled)
confusion_matrix(y_test,knnp)

knn Training scaore 0.70796460177
knn Test score 0.610619469027


array([[58,  1,  0,  1,  1,  0,  0,  0,  0,  0,  0],
       [ 8,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  4,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0],
       [ 2,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [10,  1,  0,  0,  1,  0,  0,  0,  1,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0]])

In [12]:
print("Mean Cross Validation Score of KNN:", 
      '%.4f' %cross_val_score(KNeighborsClassifier(n_neighbors=2), x_trainscaled, y_train, cv = 10).mean())

Mean Cross Validation Score of KNN: 0.5978


C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


#### KNN with Arrhythmia detection data

In [13]:
knnclas=KNeighborsClassifier()
k_value={'n_neighbors':range(1,100)}
grid=GridSearchCV(knnclas,param_grid=k_value)
grid.fit(x_trainscaled,z_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,z_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,z_test))
print(grid.best_params_)
print(grid.best_estimator_)
knnclasp=grid.predict(x_testscaled)
confusion_matrix(z_test,knnclasp)

The best training score is :  1.0000
The best test score is :  0.6372
{'n_neighbors': 1}
KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')


array([[50, 11],
       [30, 22]])

In [14]:
knna=KNeighborsClassifier(n_neighbors=1)
knn=knna.fit(x_trainscaled,z_train)
print("knn Training scaore", knn.score(x_trainscaled,z_train))
print("knn Test score",knn.score(x_testscaled,z_test))
knnp=knn.predict(x_testscaled)
confusion_matrix(z_test,knnp)

knn Training scaore 1.0
knn Test score 0.637168141593


array([[50, 11],
       [30, 22]])

In [15]:
print("Mean Cross Validation Score of KNN:", 
      '%.4f' %cross_val_score(KNeighborsClassifier(n_neighbors=1), x_trainscaled, z_train, cv = 10).mean())

Mean Cross Validation Score of KNN: 0.6723


KNN is known for its memory. It remembers whole training data, which results in maximum training accuracy. But it is performing very poorly on the test data. So, we can expect that the number training examples is not enough for KNN. Because we have only 452 observations. So, knn is ruled out of our consideration.

### Logistic Regression

#### Multiclass Data

In [16]:
lrg=LogisticRegression()
value={'penalty':['l1','l2'], 'C':[0.1,0.5,0.75,1,2,5,10,15,20,30,50,100]}
grid=GridSearchCV(lrg,param_grid=value)
grid.fit(x_trainscaled,y_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,y_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,y_test))
print(grid.best_params_)
print(grid.best_estimator_)
lrgp=grid.predict(x_testscaled)
confusion_matrix(y_test,lrgp)

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


The best training score is :  0.8496
The best test score is :  0.6814
{'C': 2, 'penalty': 'l1'}
LogisticRegression(C=2, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)


array([[56,  3,  0,  0,  0,  0,  0,  0,  2,  0,  0,  0],
       [ 7,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  2,  0,  0,  0,  1,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  1,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0,  0],
       [ 3,  1,  0,  0,  0,  0,  0,  0,  9,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  0,  0,  0,  1,  0,  0,  1],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  0,  1,  0,  0,  0,  0,  0,  0,  0,  1,  0]])

In [17]:
lrm=LogisticRegression(C=2,penalty='l1')
lr=lrm.fit(x_trainscaled,y_train)
print("Logistics Training score", lr.score(x_trainscaled,y_train))
print("Logistics Test Score",lr.score(x_testscaled,y_test))
lrp=lr.predict(x_testscaled)
confusion_matrix(y_test,lrp)

Logistics Training score 0.849557522124
Logistics Test Score 0.681415929204


array([[56,  3,  0,  0,  0,  0,  0,  0,  2,  0,  0,  0],
       [ 7,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  2,  0,  0,  0,  1,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  1,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0,  0],
       [ 3,  1,  0,  0,  0,  0,  0,  0,  9,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  0,  0,  0,  1,  0,  0,  1],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  0,  1,  0,  0,  0,  0,  0,  0,  0,  1,  0]])

In [18]:
print("Mean Cross Validation Score of Logistic Regression:", 
      '%.4f' %cross_val_score(LogisticRegression(C=0.75,penalty='l2'), x_trainscaled, y_train, cv = 10).mean())

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


Mean Cross Validation Score of Logistic Regression: 0.7063


Moving to the Logistic regression, we can observe that the testing accuracy on multiclass data is little better than KNN model, but we cannot say the best because it is clearly overfitting the data. The best model is obtained at c=0.75 with a penalty of L1

#### Arrhythmia detection data

In [19]:
lrg=LogisticRegression()
value={'penalty':['l1','l2'], 'C':[0.1,0.5,0.75,1,2,5,10,15,20,30,50,100]}
grid=GridSearchCV(lrg,param_grid=value)
grid.fit(x_trainscaled,z_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,z_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,z_test))
print(grid.best_params_)
print(grid.best_estimator_)
lrgp=grid.predict(x_testscaled)
confusion_matrix(z_test,lrgp)

The best training score is :  0.8525
The best test score is :  0.7168
{'C': 0.75, 'penalty': 'l2'}
LogisticRegression(C=0.75, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)


array([[50, 11],
       [21, 31]])

In [20]:
lra=LogisticRegression(C=0.75,penalty='l2')
lr=lra.fit(x_trainscaled,z_train)
print("Logistics Training score", lr.score(x_trainscaled,z_train))
print("Logistics Test Score",lr.score(x_testscaled,z_test))
lrp=lr.predict(x_testscaled)
confusion_matrix(z_test,lrp)

Logistics Training score 0.852507374631
Logistics Test Score 0.716814159292


array([[50, 11],
       [21, 31]])

In [21]:
print("Mean Cross Validation Score of Logistic Regression:", 
      '%.4f' %cross_val_score(LogisticRegression(C=0.75,penalty='l2'), x_trainscaled, z_train, cv = 10).mean())

Mean Cross Validation Score of Logistic Regression: 0.7904


Comparing to the Multiclass data, we can say that logistic regression is performing better in detection of Arrhythmia resulting a cross validation score close to 80%.

### Linear Support Vector Machines

#### Multiclass Data

In [22]:
svcg=LinearSVC()
value={'C':[0.1,0.2,0.3,0.4,0.5,0.6,0.75,1,2,5,10,15,20,30,50,100]}
grid=GridSearchCV(svcg,param_grid=value)
grid.fit(x_trainscaled,y_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,y_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,y_test))
print(grid.best_params_)
print(grid.best_estimator_)
svcgp=grid.predict(x_testscaled)
confusion_matrix(y_test,svcgp)

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


The best training score is :  0.8909
The best test score is :  0.6814
{'C': 0.3}
LinearSVC(C=0.3, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)


array([[56,  3,  0,  0,  0,  0,  0,  0,  2,  0,  0,  0],
       [ 5,  2,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1],
       [ 0,  1,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  2,  0,  0,  0,  1,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  0,  0,  0,  0,  0,  0,  0,  9,  0,  0,  0],
       [ 1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 3,  0,  1,  0,  0,  1,  0,  0,  0,  0,  1,  0]])

In [23]:
svclm=LinearSVC(C=0.3)
svc=svclm.fit(x_trainscaled,y_train)
print("The Best Training Score",svc.score(x_trainscaled,y_train))
print("The Best Test Score",svc.score(x_testscaled,y_test))
svcp=svc.predict(x_testscaled)
confusion_matrix(y_test,svcp)

The Best Training Score 0.890855457227
The Best Test Score 0.681415929204


array([[56,  3,  0,  0,  0,  0,  0,  0,  2,  0,  0,  0],
       [ 5,  2,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1],
       [ 0,  1,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  2,  0,  0,  0,  1,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  0,  0,  0,  0,  0,  0,  0,  9,  0,  0,  0],
       [ 1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 3,  0,  1,  0,  0,  1,  0,  0,  0,  0,  1,  0]])

In [24]:
print("Mean Cross Validation Score of Linear SVC:", 
      '%.4f' %cross_val_score(LinearSVC(C=0.3), x_trainscaled, y_train, cv = 10).mean())

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


Mean Cross Validation Score of Linear SVC: 0.7298


#### Arrhythmia detection data

In [25]:
svcg=LinearSVC()
value={'C':[0.05,0.1,0.15,0.4,0.5,0.75,1,2,5,10,15,20,30,50,100]}
grid=GridSearchCV(svcg,param_grid=value)
grid.fit(x_trainscaled,z_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,z_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,z_test))
print(grid.best_params_)
print(grid.best_estimator_)
svcgp=grid.predict(x_testscaled)
confusion_matrix(z_test,svcgp)

The best training score is :  0.8643
The best test score is :  0.7257
{'C': 0.15}
LinearSVC(C=0.15, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)


array([[50, 11],
       [20, 32]])

In [26]:
svcla=LinearSVC(C=0.15)
svc=svcla.fit(x_trainscaled,z_train)
print("The best Training score is :", svc.score(x_trainscaled,z_train))
print("The best Test Score is :",svc.score(x_testscaled,z_test))
svcp=svc.predict(x_testscaled)
confusion_matrix(z_test,svcp)

The best Training score is : 0.864306784661
The best Test Score is : 0.725663716814


array([[50, 11],
       [20, 32]])

In [27]:
print("Mean Cross Validation Score of Linear SVC:", 
      '%.4f' %cross_val_score(LinearSVC(C=0.15), x_trainscaled, z_train, cv = 10).mean())

Mean Cross Validation Score of Linear SVC: 0.7756


Linear support vector machines is kind of overfitting in both cases but comparitively it is performing best in identifying the Arrhythmia. But we cannot say that it is the best, but still standing in the race with a good cross validation score of 77.56%

## Kerenilzed Support Vector Machine

### RBF Kernel

#### MultiClass Data

In [28]:
svcg=SVC(kernel='rbf')
value = {'C':[0.1,1,5,10,12,15,17,20,25], 'gamma':[0.01,0.05,0.1,0.25,0.5, 1, 5]}
grid=GridSearchCV(svcg,param_grid=value)
grid.fit(x_trainscaled,y_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,y_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,y_test))
print(grid.best_params_)
print(grid.best_estimator_)
svcgp=grid.predict(x_testscaled)
confusion_matrix(y_test,svcgp)

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


The best training score is :  0.8024
The best test score is :  0.6726
{'C': 12, 'gamma': 0.01}
SVC(C=12, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)


array([[58,  2,  0,  0,  0,  0,  0,  0,  1,  0,  0],
       [ 6,  2,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  5,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0],
       [ 2,  0,  0,  0,  1,  0,  0,  0,  1,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  7,  0,  0],
       [ 2,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0]])

In [29]:
svcrm=SVC(kernel='rbf',C=12,gamma=0.01)
svc=svcrm.fit(x_trainscaled,y_train)
print("The best Training score is :", svc.score(x_trainscaled,y_train))
print("The best Test Score is :",svc.score(x_testscaled,y_test))
svcp=svc.predict(x_testscaled)
confusion_matrix(y_test,svcp)

The best Training score is : 0.802359882006
The best Test Score is : 0.672566371681


array([[58,  2,  0,  0,  0,  0,  0,  0,  1,  0,  0],
       [ 6,  2,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  5,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0],
       [ 2,  0,  0,  0,  1,  0,  0,  0,  1,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  7,  0,  0],
       [ 2,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0]])

In [30]:
print("Mean Cross Validation Score of Kernalized SVC:", 
      '%.4f' %cross_val_score(SVC(kernel='rbf',C=12,gamma=0.01), x_trainscaled, y_train, cv = 10).mean())

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


Mean Cross Validation Score of Kernalized SVC: 0.7144


#### Arrhythmia detection data

In [31]:
svcg=SVC(kernel='rbf')
value = {'C':[0.1,1,3,5,7,10], 'gamma':[0.01,0.05,0.1,0.25,0.5, 1, 5]}
grid=GridSearchCV(svcg,param_grid=value)
grid.fit(x_trainscaled,z_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,z_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,z_test))
print(grid.best_params_)
print(grid.best_estimator_)
svcgp=grid.predict(x_testscaled)
confusion_matrix(z_test,svcgp)

The best training score is :  0.8938
The best test score is :  0.7522
{'C': 5, 'gamma': 0.05}
SVC(C=5, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.05, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)


array([[51, 10],
       [18, 34]])

In [32]:
svcra=SVC(kernel='rbf',C=5,gamma=0.05)
svc=svcra.fit(x_trainscaled,z_train)
print("The best Training score is :", svc.score(x_trainscaled,z_train))
print("The best Test Score is :",svc.score(x_testscaled,z_test))
svcp=svc.predict(x_testscaled)
confusion_matrix(z_test,svcp)

The best Training score is : 0.893805309735
The best Test Score is : 0.752212389381


array([[51, 10],
       [18, 34]])

In [33]:
print("Mean Cross Validation Score of Kernalized SVC:", 
      '%.4f' %cross_val_score(SVC(kernel='rbf',C=5,gamma=0.05), x_trainscaled, z_train, cv = 10).mean())

Mean Cross Validation Score of Kernalized SVC: 0.7818


### Polynomial Kernel

#### Multinomial Data

In [34]:
svcg=SVC(kernel='poly')
value = {'degree':[2,3,4,5],'C':[0.1,0.5,1,2,5,10], 'gamma':[0.01,0.05,0.1,0.25,0.5, 1, 5]}
grid=GridSearchCV(svcg,param_grid=value)
grid.fit(x_trainscaled,y_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,y_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,y_test))
print(grid.best_params_)
print(grid.best_estimator_)
svcgp=grid.predict(x_testscaled)
confusion_matrix(y_test,svcgp)

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


The best training score is :  0.8142
The best test score is :  0.6903
{'C': 1, 'degree': 2, 'gamma': 0.05}
SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=2, gamma=0.05, kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)


array([[58,  2,  0,  0,  0,  0,  0,  0,  1,  0,  0],
       [ 6,  2,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  5,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  2,  0,  0,  0,  1,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  0,  0,  0,  8,  0,  0],
       [ 2,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0]])

In [35]:
svcpm=SVC(kernel='poly',degree=2,C=1,gamma=0.05)
svc=svcpm.fit(x_trainscaled,y_train)
print("The best Training score is :", svc.score(x_trainscaled,y_train))
print("The best Test Score is :",svc.score(x_testscaled,y_test))
svcp=svc.predict(x_testscaled)
confusion_matrix(y_test,svcp)

The best Training score is : 0.814159292035
The best Test Score is : 0.690265486726


array([[58,  2,  0,  0,  0,  0,  0,  0,  1,  0,  0],
       [ 6,  2,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  5,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  2,  0,  0,  0,  1,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  0,  0,  0,  8,  0,  0],
       [ 2,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0]])

In [36]:
print("Mean Cross Validation Score of Kernalized SVC:", 
      '%.4f' %cross_val_score(SVC(kernel='poly',degree=2,C=1,gamma=0.05), x_trainscaled, y_train, cv = 10).mean())

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


Mean Cross Validation Score of Kernalized SVC: 0.7141


#### Arrhythmia detection Data

In [37]:
svcg=SVC(kernel='poly')
value = {'degree':[2,3,4,5],'C':[0.1,0.2,0.3,0.4,0.5,0.6,0.7,1,2,5,10], 'gamma':[0.01,0.05,0.1,0.25,0.5, 1, 5]}
grid=GridSearchCV(svcg,param_grid=value)
grid.fit(x_trainscaled,z_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,z_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,z_test))
print(grid.best_params_)
print(grid.best_estimator_)
svcgp=grid.predict(x_testscaled)
confusion_matrix(z_test,svcgp)

The best training score is :  0.8761
The best test score is :  0.7168
{'C': 0.5, 'degree': 3, 'gamma': 0.05}
SVC(C=0.5, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.05, kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)


array([[49, 12],
       [20, 32]])

In [38]:
svcpa=SVC(kernel='poly',degree=3,C=0.5,gamma=0.05)
svc=svcpa.fit(x_trainscaled,z_train)
print("The best Training score is :", svc.score(x_trainscaled,z_train))
print("The best Test Score is :",svc.score(x_testscaled,z_test))
svcp=svc.predict(x_testscaled)
confusion_matrix(z_test,svcp)

The best Training score is : 0.87610619469
The best Test Score is : 0.716814159292


array([[49, 12],
       [20, 32]])

In [39]:
print("Mean Cross Validation Score of Kernalized SVC:", 
      '%.4f' %cross_val_score(SVC(kernel='poly',degree=3,C=0.5,gamma=0.05), x_trainscaled, z_train, cv = 10).mean())

Mean Cross Validation Score of Kernalized SVC: 0.7642


There is not much improvement in the kernelized support vector in both RBF and Polynomial kernel. Both the cases are overfitting in classifying multi-class data as well as detectin Arrhythmia. But the cross validation scores are satisfactory.

### Decision Tree

#### Multiclass Data

In [40]:
dt=DecisionTreeClassifier()
value={"max_depth":range(1,20)}
grid=GridSearchCV(dt,param_grid=value)
grid.fit(x_trainscaled,y_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,y_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,y_test))
print(grid.best_params_)
print(grid.best_estimator_)
dtp=grid.predict(x_testscaled)
confusion_matrix(y_test,dtp)

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


The best training score is :  0.8555
The best test score is :  0.6018
{'max_depth': 6}
DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=6,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')


array([[51,  4,  1,  1,  0,  1,  0,  0,  3,  0,  0],
       [ 4,  3,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [ 1,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 2,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  1,  0,  0,  0,  1,  0,  1],
       [ 0,  0,  0,  0,  0,  5,  0,  0,  1,  0,  0],
       [ 0,  1,  0,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  1,  1,  0,  0,  0,  0,  0,  7,  0,  0],
       [ 1,  0,  0,  0,  0,  0,  0,  0,  1,  0,  1],
       [ 4,  1,  0,  0,  0,  1,  0,  0,  0,  0,  0]])

In [41]:
dtm=DecisionTreeClassifier(max_depth = 6)
dt=dtm.fit(x_trainscaled,y_train)
print("The best Training score is :", dt.score(x_trainscaled,y_train))
print("The best Test Score is :",dt.score(x_testscaled,y_test))
dtp=dt.predict(x_testscaled)
confusion_matrix(y_test,dtp)

The best Training score is : 0.855457227139
The best Test Score is : 0.601769911504


array([[51,  5,  1,  0,  1,  1,  0,  0,  2,  0,  0,  0],
       [ 5,  3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 2,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  1,  0,  0,  1,  0,  0,  0,  0,  0,  0,  1],
       [ 1,  0,  0,  0,  0,  5,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  1,  1,  0,  0,  0,  0,  0,  7,  0,  0,  0],
       [ 1,  0,  0,  0,  0,  0,  0,  0,  1,  0,  1,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 4,  1,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0]])

In [42]:
print("Mean Cross Validation Score of Decision Tree is :", 
      '%.4f' %cross_val_score(DecisionTreeClassifier(max_depth = 6), x_trainscaled, y_train, cv = 10).mean())

Mean Cross Validation Score of Decision Tree is : 0.6955


C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


#### Arrhythmia detection Data

In [43]:
dt=DecisionTreeClassifier()
value={"max_depth":range(1,20)}
grid=GridSearchCV(dt,param_grid=value)
grid.fit(x_trainscaled,z_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,z_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,z_test))
print(grid.best_params_)
print(grid.best_estimator_)
dtp=grid.predict(x_testscaled)
confusion_matrix(z_test,dtp)

The best training score is :  0.9027
The best test score is :  0.7257
{'max_depth': 5}
DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=5,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')


array([[48, 13],
       [18, 34]])

In [44]:
dta=DecisionTreeClassifier(max_depth = 6)
dt=dta.fit(x_trainscaled,z_train)
print("The best Training score is :", dt.score(x_trainscaled,z_train))
print("The best Test Score is :",dt.score(x_testscaled,z_test))
dtp=dt.predict(x_testscaled)
confusion_matrix(z_test,dtp)

The best Training score is : 0.93215339233
The best Test Score is : 0.699115044248


array([[43, 18],
       [16, 36]])

In [45]:
print("Mean Cross Validation Score of Decision Tree is :", 
      '%.4f' %cross_val_score(DecisionTreeClassifier(max_depth = 6), x_trainscaled, z_train, cv = 10).mean())

Mean Cross Validation Score of Decision Tree is : 0.7604


Decision Trees are performing well in dsetection of Arrhythmia but failing to detect the exact reason for Arrhythmia and the best scores are observed at a tree depth of 6.

### Random Forest

#### Multiclass Data

In [46]:
rf = RandomForestClassifier(random_state = 0)
value={"n_estimators":range(1,40)}
grid=GridSearchCV(rf,param_grid=value)
grid.fit(x_trainscaled,y_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,y_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,y_test))
print(grid.best_params_)
print(grid.best_estimator_)
rfp=grid.predict(x_testscaled)
confusion_matrix(y_test,rfp)

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


The best training score is :  0.9971
The best test score is :  0.6726
{'n_estimators': 35}
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            n_estimators=35, n_jobs=1, oob_score=False, random_state=0,
            verbose=0, warm_start=False)


array([[55,  2,  0,  0,  0,  1,  0,  0,  3,  0,  0],
       [ 5,  3,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  5,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  1,  0,  0,  0,  2,  0,  0],
       [ 4,  0,  0,  0,  0,  2,  0,  0,  0,  0,  0],
       [ 0,  0,  1,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [ 4,  0,  0,  0,  0,  0,  0,  0,  9,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0]])

In [47]:
rfm = RandomForestClassifier(n_estimators=35,random_state = 0)
rf=rfm.fit(x_trainscaled,y_train)
print("The best Training score is :", rf.score(x_trainscaled,y_train))
print("The best Test Score is :",rf.score(x_testscaled,y_test))
rfp=rf.predict(x_testscaled)
confusion_matrix(y_test,rfp)

The best Training score is : 0.997050147493
The best Test Score is : 0.672566371681


array([[55,  2,  0,  0,  0,  1,  0,  0,  3,  0,  0],
       [ 5,  3,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  5,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  1,  0,  0,  0,  2,  0,  0],
       [ 4,  0,  0,  0,  0,  2,  0,  0,  0,  0,  0],
       [ 0,  0,  1,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],
       [ 4,  0,  0,  0,  0,  0,  0,  0,  9,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 6,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0]])

In [48]:
print("Mean Cross Validation Score of Random Forest is :", 
      '%.4f' %cross_val_score(RandomForestClassifier(n_estimators=35,random_state = 0), x_trainscaled, y_train, cv = 10).mean())

C:\Users\satya\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:581: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of groups for any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


Mean Cross Validation Score of Random Forest is : 0.7340


#### Arrhythmia detection Data

In [49]:
rf = RandomForestClassifier(random_state = 0)
value={"n_estimators":range(1,40)}
grid=GridSearchCV(rf,param_grid=value)
grid.fit(x_trainscaled,z_train)
print("The best training score is : ",'%.4f' %grid.score(x_trainscaled,z_train))
print("The best test score is : ",'%.4f' %grid.score(x_testscaled,z_test))
print(grid.best_params_)
print(grid.best_estimator_)
rfp=grid.predict(x_testscaled)
confusion_matrix(z_test,rfp)

The best training score is :  1.0000
The best test score is :  0.7257
{'n_estimators': 30}
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            n_estimators=30, n_jobs=1, oob_score=False, random_state=0,
            verbose=0, warm_start=False)


array([[47, 14],
       [17, 35]])

In [50]:
rfa = RandomForestClassifier(n_estimators=30,random_state = 0)
rf=rfa.fit(x_trainscaled,z_train)
print("The best Training score is :", rf.score(x_trainscaled,z_train))
print("The best Test Score is :",rf.score(x_testscaled,z_test))
rfp=rf.predict(x_testscaled)
confusion_matrix(z_test,rfp)

The best Training score is : 1.0
The best Test Score is : 0.725663716814


array([[47, 14],
       [17, 35]])

In [51]:
print("Mean Cross Validation Score of Random Forest is :", 
      '%.4f' %cross_val_score(RandomForestClassifier(n_estimators=30,random_state = 0), x_trainscaled, z_train, cv = 10).mean())

Mean Cross Validation Score of Random Forest is : 0.8088


Finally the Random Forest came into the race with highest cross validation score of 81% in detecting the Arrhythmia. This is obtained with the 30 estimators with a random state of 0

### Bagging

Bootstrap aggregating, also called bagging, is a machine learning ensemble meta-algorithm designed to improve the stability and accuracy of machine learning algorithms used in statistical classification and regression. It also reduces variance and helps to avoid overfitting

#### Multi-Class Data

In [52]:
models = [knnm,lrm,svclm,svcrm,svcpm,dtm,rfm]

from sklearn.ensemble import BaggingClassifier

for model in models:
    bag = BaggingClassifier(model,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
    bag.fit(x_trainscaled,y_train)
    print('model:',model)
    print('training score:',bag.score(x_trainscaled,y_train))
    print('testing score:',bag.score(x_testscaled,y_test))
    print("\n")

model: KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=2, p=2,
           weights='uniform')
training score: 0.619469026549
testing score: 0.592920353982


model: LogisticRegression(C=2, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
training score: 0.705014749263
testing score: 0.646017699115


model: LinearSVC(C=0.3, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)
training score: 0.890855457227
testing score: 0.681415929204


model: SVC(C=12, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, pr

By applying bagging on multiclass data, we can observe that there is a lot of improvement in training the model. It clearly tried hard to reduce the overfitting but failed to obtain the best model.
The only model that got benefitted from bagging is Decision Trees, where you can observe a clear increaase of 10% in testing accuracy.

#### Arrhythmia detection Data

In [53]:
models = [knnm,lrm,svclm,svcrm,svcpm,dtm,rfm]

from sklearn.ensemble import BaggingClassifier

for model in models:
    bag = BaggingClassifier(model,n_estimators=100,max_samples=100,bootstrap=True,n_jobs=-1,random_state=10)
    bag.fit(x_trainscaled,y_train)
    print('model:',model)
    print('training score:',bag.score(x_trainscaled,y_train))
    print('testing score:',bag.score(x_testscaled,y_test))
    print("\n")

model: KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=2, p=2,
           weights='uniform')
training score: 0.619469026549
testing score: 0.592920353982


model: LogisticRegression(C=2, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
training score: 0.705014749263
testing score: 0.646017699115


model: LinearSVC(C=0.3, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)
training score: 0.890855457227
testing score: 0.681415929204


model: SVC(C=12, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, pr

Only Decision Trees got benefitted by applying bagging.

### AdaBoostClassifier

#### Multi-Class Data

In [54]:
models = [lrm,svclm,svcrm,svcpm,dtm,rfm]

from sklearn.ensemble import AdaBoostClassifier

for model in models:
    ada = AdaBoostClassifier(model,n_estimators=100,algorithm='SAMME',learning_rate=0.5,random_state=42)
    ada.fit(x_trainscaled,y_train)
    print('model:',model)
    print('training score:',ada.score(x_trainscaled,y_train))
    print('testing score:', ada.score(x_testscaled,y_test))
    print("\n")

model: LogisticRegression(C=2, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
training score: 0.542772861357
testing score: 0.53982300885


model: LinearSVC(C=0.3, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)
training score: 0.890855457227
testing score: 0.681415929204


model: SVC(C=12, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
training score: 0.542772861357
testing score: 0.53982300885


model: SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None,

#### Arrhythmia detection Data

In [55]:
models = [lrm,svclm,svcrm,svcpm,dtm,rfm]

from sklearn.ensemble import AdaBoostClassifier

for model in models:
    ada = AdaBoostClassifier(model,n_estimators=100,algorithm='SAMME',learning_rate=0.5,random_state=42)
    ada.fit(x_trainscaled,z_train)
    print('model:',model)
    print('training score:',ada.score(x_trainscaled,z_train))
    print('testing score:', ada.score(x_testscaled,z_test))
    print("\n")

model: LogisticRegression(C=2, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
training score: 0.542772861357
testing score: 0.53982300885


model: LinearSVC(C=0.3, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)
training score: 0.896755162242
testing score: 0.743362831858


model: SVC(C=12, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
training score: 0.542772861357
testing score: 0.53982300885


model: SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None,

Boosting is a general ensemble method that creates a strong classifier from a number of weak classifiers.
This is done by building a model from the training data, then creating a second model that attempts to correct the errors from the first model. Models are added until the training set is predicted perfectly or a maximum number of models are added.
AdaBoost is best used to boost the performance of decision trees on binary classification problems.

We can clearly observe that Adaboost failed in improving the prediction on multiclass data, but huge success in improving the binary data using the decision trees. 

The performance of decision tree increased by 20%.

### GradientBoostingClassifier

#### Multi-Class Data

In [56]:
from sklearn.ensemble import GradientBoostingClassifier

for depth in range(1,10):
    for rate in (0.001,0.01,0.1):
        gbModel = GradientBoostingClassifier(max_depth=depth,learning_rate=rate,random_state=0)
        gbModel.fit(x_trainscaled,y_train)
        print('depth:',depth)
        print('learning rate:',rate)
        print('training score:',gbModel.score(x_trainscaled,y_train))
        print('testing score:',gbModel.score(x_testscaled,y_test))
        print("\n")

depth: 1
learning rate: 0.001
training score: 0.663716814159
testing score: 0.592920353982


depth: 1
learning rate: 0.01
training score: 0.843657817109
testing score: 0.699115044248


depth: 1
learning rate: 0.1
training score: 0.979351032448
testing score: 0.699115044248


depth: 2
learning rate: 0.001
training score: 0.749262536873
testing score: 0.601769911504


depth: 2
learning rate: 0.01
training score: 0.914454277286
testing score: 0.716814159292


depth: 2
learning rate: 0.1
training score: 1.0
testing score: 0.725663716814


depth: 3
learning rate: 0.001
training score: 0.834808259587
testing score: 0.628318584071


depth: 3
learning rate: 0.01
training score: 0.952802359882
testing score: 0.681415929204


depth: 3
learning rate: 0.1
training score: 1.0
testing score: 0.699115044248


depth: 4
learning rate: 0.001
training score: 0.87610619469
testing score: 0.663716814159


depth: 4
learning rate: 0.01
training score: 0.979351032448
testing score: 0.690265486726


depth: 4
l

Gradient Boosting increased the performance of model on multiclass data by 15% at a depth of 5 with a learning rate of 0.1. But still it is overfitting because the training error is 0 but the testing error is still 25% even it is the best model sofar.

#### Arrhythmia detection Data

In [57]:
from sklearn.ensemble import GradientBoostingClassifier

for depth in range(1,10):
    for rate in (0.001,0.01,0.1):
        gbModel = GradientBoostingClassifier(max_depth=depth,learning_rate=rate,random_state=0)
        gbModel.fit(x_trainscaled,z_train)
        print('depth:',depth)
        print('learning rate:',rate)
        print('training score:',gbModel.score(x_trainscaled,z_train))
        print('testing score:',gbModel.score(x_testscaled,z_test))
        print("\n")

depth: 1
learning rate: 0.001
training score: 0.542772861357
testing score: 0.53982300885


depth: 1
learning rate: 0.01
training score: 0.784660766962
testing score: 0.699115044248


depth: 1
learning rate: 0.1
training score: 0.920353982301
testing score: 0.823008849558


depth: 2
learning rate: 0.001
training score: 0.678466076696
testing score: 0.628318584071


depth: 2
learning rate: 0.01
training score: 0.867256637168
testing score: 0.752212389381


depth: 2
learning rate: 0.1
training score: 0.982300884956
testing score: 0.83185840708


depth: 3
learning rate: 0.001
training score: 0.722713864307
testing score: 0.619469026549


depth: 3
learning rate: 0.01
training score: 0.923303834808
testing score: 0.769911504425


depth: 3
learning rate: 0.1
training score: 0.997050147493
testing score: 0.83185840708


depth: 4
learning rate: 0.001
training score: 0.793510324484
testing score: 0.672566371681


depth: 4
learning rate: 0.01
training score: 0.946902654867
testing score: 0.78761

In [58]:
gbModel = GradientBoostingClassifier(max_depth=1,learning_rate=0.1,random_state=0)
gbModel.fit(x_trainscaled,z_train)
print('training score:',gbModel.score(x_trainscaled,z_train))
print('testing score:',gbModel.score(x_testscaled,z_test))

training score: 0.920353982301
testing score: 0.823008849558


Gradient Boosting is doing awesome job at a depth of 3 with a learning rate of 0.1 resulting a 25% increase in testing accuracy and even the model looks good as the training error is 8% and the testing error is 14%. 
So, We can say that the gradient boosting improving the results by a great extend and standing as the best model.

### PCA

Principle component analysis is mainly used for dimentionality reduction by preserving the varience in trhe dataset. Here we have 270 feature which are making the model to overfit. So, we try to reduce the number of features to 50 by applying the principle component analysis on the data.

In [67]:
from sklearn.decomposition import PCA

pca = PCA(n_components = 50)
datapca = pca.fit_transform(x)
print(pca.explained_variance_ratio_)
datapca.shape

[ 0.15346672  0.10006548  0.09420965  0.0845956   0.05312036  0.05062328
  0.04585218  0.04113792  0.03395551  0.0291003   0.02830696  0.02249531
  0.02190761  0.01932186  0.01731159  0.01539366  0.01445182  0.01225009
  0.01176798  0.00914527  0.00867875  0.00766778  0.00722819  0.00669843
  0.0063108   0.00584685  0.0054186   0.00527819  0.0049771   0.00478087
  0.00445785  0.00419111  0.00401413  0.00373631  0.00351611  0.00327154
  0.00305062  0.00299937  0.00286984  0.00253784  0.00243258  0.00230367
  0.00220401  0.00216197  0.00209371  0.00198413  0.00195335  0.00172184
  0.00159579  0.00145356]


(452, 50)

In [68]:
pca_train,pca_test,y_train,y_test=train_test_split(datapca,y,random_state=0)

models = [lrm,svclm,svcrm,svcpm,dtm,rfm]
for i in range(5):
    models[i].fit(pca_train,y_train)
    print('model:',models[i])
    print('training score:',models[i].score(pca_train,y_train))
    print('testing score:',models[i].score(pca_test,y_test))
    print("\n")

model: LogisticRegression(C=2, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
training score: 0.917404129794
testing score: 0.575221238938


model: LinearSVC(C=0.3, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)
training score: 0.672566371681
testing score: 0.486725663717


model: SVC(C=12, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
training score: 1.0
testing score: 0.53982300885


model: SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=2,

In [69]:
pca_train,pca_test,z_train,z_test=train_test_split(datapca,z,random_state=0)

models = [lra,svcla,svcra,svcpa,dta,rfa]
for i in range(5):
    models[i].fit(pca_train,z_train)
    print('model:',models[i])
    print('training score:',models[i].score(pca_train,z_train))
    print('testing score:',models[i].score(pca_test,z_test))
    print("\n")

model: LogisticRegression(C=0.75, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
training score: 0.870206489676
testing score: 0.743362831858


model: LinearSVC(C=0.15, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)
training score: 0.787610619469
testing score: 0.654867256637


model: SVC(C=5, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.05, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
training score: 1.0
testing score: 0.53982300885


model: SVC(C=0.5, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degr

By applying Principle component Analysis, we observed that there is total reduce in the performance in all the model. Main reason for this is decrease in number of features in training the model results in failing the model in capturing the original target label.

### Observation

1. By applyting different models on the data, we observed the variations in training and testing scores across different model. Random Forest is the best model among all the model with a Cross Validation score of 81%.
2. Bagging succeded in reducing the overfit of different models but completely failed in improving the performance of model. The only model that got benefitted is Decision Tree with an increase in testing accuracy of 10%.
3. Gradinet boosting done a great job in improving the performance of the decision tree model and got highest accuracy of 84% on testing data, training accuracy of 92% resulting in a good fit.
4. After applying Principle component analysis, we reduced the number of features to 50 by preserving more than 95% of its variance. But we already know that by reducing the number of features, we loose information resulting in insufficient data. Because of this there is huge decrese in the accurcy levels across all the models.